### ISSUE TITLE
Setting the document font in Preferences overrides Build Settings

### ISSUE THREAD

**Issue by HeyMyian** [vkbo/novelWriter#2118]
This in 2.6 Beta 1, AppImage on KDE Neon.

**Reproduce:**
1. Change the "Document font" in the novelWriter Preferences
2. Change the "Text Font" in the MBS to something else

**Result:**
- The font from the novelWriter Preferences is used for the build preview, print preview, and for the pdf, ignoring the font from the MBS.

**Workaround:**
- Restarting novelWriter uses the MBS font for the preview and pdf.

**Comment by vkbo** [comment]
Is this behaviour present in 2.5.x as well?

**Comment by HeyMyian** [comment]
No, it works as expected in 2.5.2.

**Comment by HeyMyian** [comment]
It's present in 2.5.2 as well, although it's "Application font", not "Document font", which triggers it. This may be expected behavior though? as changing the Application font tells the user to restart novelWriter.

**Comment by vkbo** [comment]
The document font is the default font for the build preview. But it should override with the one selected in the settings. I'll have a look at what's happening here.

I was hoping to do the 2.5.3 release tomorrow, but if it's present there as well, it may be delayed. I also want to do the 2.6 Beta 2 release soon.

**Comment by vkbo** [comment]
I think this must be another Qt bug.

No matter what I do, if you change the document font, the preview and PDF output will change the font as well. I tested every single internal font value i can pull from these documents, and it does not change. It is still what's selected in the build settings. Yet it outputs the selected document font. I've evn tried setting it on the character level, and even that doesn't help.

I think there must be something in the render engine for text that simply overrides it.

Maybe it's fixed in Qt6. I will be moving to Qt6 early next year.

**Comment by HeyMyian** [comment]
It's weird. I hope I did these tests right, they're rather finicky to execute. This is on Windows. Does the setting "Use the system's font selection dialog" btw have a possible effect on these outcomes? My tests were conducted with this setting **off**.

## 2.5.2
* Setting the **Application font** overrides the Build Settings and puts the Application font name into the input field for Text Font in the MBS. Changing the MBS font is **not** possible. The preview will display the Application font.

* Setting the **Document font** has no effect on the Build Settings, as is the expected behavior.

## 2.6 Beta 1
* Setting the **Application font** has **no effect** on the Build Settings. In the MBS, the input field for Text Font is replaced with the Application font name, but that seems to me only a display error. The preview uses the previous MBS font. So this works perfectly fine, sans the display error in the input field. Changing the MBS font **is possible**, and the preview reflects that.

* Setting the **Document font** overrides the Build Settings. Changing the MBS font is **not** possible without restarting novelWriter.

---

All of this is only an issue when the Application/Document font is changed in the current novelWriter session. After a restart, the MBS font can be changed.

(I typed all of that up just before your previous message. Leaving it here in case it helps.)

**Comment by vkbo** [comment]
> It's weird. I hope I did these tests right, they're rather finicky to execute. This is on Windows. Does the setting "Use the system's font selection dialog" btw have a possible effect on these outcomes? My tests were conducted with this setting **off**.

Possibly. I know that the Gnome 3 selector is very limited, so it's better to use the Qt selector on Gnome Linux.

**Comment by vkbo** [comment]
> (I typed all of that up just before your previous message. Leaving it here in case it helps.)

Yeah, I remembered a few issues I discovered when I implemented the new font selector features. Basically, QFont is the font you've selected. But it may be a complete illusion. QFontInfo contains the actual font info of the loaded font. This is where I spotted the discrepancy when new fonts have been selected in Preferences.

Basically, changing the main font resets the font for the whole application. The document fonts are then reset afterwards, so the whole flow there is messy.

**Comment by HeyMyian** [comment]
Can you track what changed between 2.5.2 and 2.6 Beta 1 that switches the problem trigger from `Application font` to `Document font`?

If it's not solvable then maybe a notification to restart novelWriter after changing the `Document font` would be a good fix.

**Comment by vkbo** [comment]
It seems to be solvable, as long as the selected font is well defined. I created a helper function that checks that the font family you requested is what you got. If it doesn't match, it will do a manual lookup in the font database and try to find it. When it does so, it will print this in the log output:

![image](https://github.com/user-attachments/assets/91ba9a1d-c965-47ae-beeb-078cff1e8c1f)

I'm making the helper function fall back to the font Qt picked if it cannot find a better one, but it's worked so far for the cases I've tested.

**Comment by vkbo** [comment]
Ok, this seems to work. I will merge it tomorrow. I just need to fix another issue first. Maybe you can test it from the branch? If not, I will make the Beta 2 release soon.

Interestingly, if I use the Gnome 3 default font selector, none of these issues occur. But then, I can also not select font style. I get "Regular" regardless of what I actually select.

```python
def fontMatcher(font: QFont) -> QFont:
    """Make sure the font is the correct family, if possible. This
    ensures that Qt doesn't use the GUI or document font instead.
    """
    info = QFontInfo(font)
    if (famRequest := font.family()) != (famActual := info.family()):
        logger.warning("Font mismatch: %s != %s", famRequest, famActual)
        db = QFontDatabase()
        if famRequest in db.families():
            styleRequest, sizeRequest = font.styleName(), font.pointSize()
            logger.info("Lookup: %s, %s, %d pt", famRequest, styleRequest, sizeRequest)
            temp = db.font(famRequest, styleRequest, sizeRequest)
            famFound, styleFound, sizeFound = temp.family(), temp.styleName(), temp.pointSize()
            if famFound == famRequest:
                logger.info("Found: %s, %s, %d pt", famFound, styleFound, sizeFound)
                return temp
        logger.warning("Could not find a font macth")
        logger.warning("You may need to restart the application")
    return font
```

**Comment by HeyMyian** [comment]
> Maybe you can test it from the branch?

The issue as described in this ticked is fixed ✅ 

After changing "Application font", the wrong font name will still be displayed in the Build Settings. But this is only a display error and is probably covered by the advice to restart novelWriter after changing the Application font.

Testing in the font_styles branch, I randomly selected "Courier" as font in the MBS. I also have "Courier New" installed. Is it possible that there is some interference with the names of these font families? Because selecting "Courier" creates only pdfs with a very tiny font size, no matter which font size I chose in the MBS. The build preview however is fine. In 2.6 Beta 1, this issue does not occur. All other fonts in the branch seem fine though.

**Comment by vkbo** [comment]
Could you run novelWriter with `--debug` or `--info`? Then you'll get the log output above. It should tell you what font it picks. You can also add `--color` now so the warning is easier to spot. (Not sure if this works on Windows, but it does on Linux.)

**Comment by HeyMyian** [comment]
I'm not seeing any info about fonts in the console with `--debug` 😦 Am I looking in the wrong place?
This is me setting the font to "Courier 32pt" in the Build Settings:
```
[2024-11-26 02:42:34,920]    manussettings.py:80    DEBUG     Create: GuiBuildSettings
[2024-11-26 02:42:34,933]    manussettings.py:155   DEBUG     Ready: GuiBuildSettings
[2024-11-26 02:42:34,952]          guimain.py:964   DEBUG     Main focus switched to: QLineEdit
[2024-11-26 02:42:34,965]    manussettings.py:419   DEBUG     Building project tree
[2024-11-26 02:42:38,632]          guimain.py:964   DEBUG     Main focus switched to: _FormattingTab
[2024-11-26 02:42:38,738]          guimain.py:964   DEBUG     Main focus switched to: QListView
[2024-11-26 02:42:40,228]          guimain.py:964   DEBUG     Main focus switched to: QListView
[2024-11-26 02:42:40,847]          guimain.py:964   DEBUG     Main focus switched to: QPushButton
[2024-11-26 02:42:40,896]          guimain.py:964   DEBUG     Main focus switched to: GuiDocEditor
[2024-11-26 02:42:41,457]          guimain.py:964   DEBUG     Main focus switched to: _FormattingTab
[2024-11-26 02:42:41,457]          guimain.py:964   DEBUG     Main focus switched to: QPushButton
[2024-11-26 02:42:41,515]    buildsettings.py:646   DEBUG     Saving builds file
[2024-11-26 02:42:41,515]    buildsettings.py:455   DEBUG     Collecting build setting for 'Mein Manuskript'
[2024-11-26 02:42:41,517]    manussettings.py:188   DEBUG     Closing: GuiBuildSettings
[2024-11-26 02:42:41,517]    manussettings.py:247   DEBUG     Saving State: GuiBuildSettings
[2024-11-26 02:42:41,519]          options.py:133   DEBUG     Saving GUI options file
[2024-11-26 02:42:41,523]    manussettings.py:160   DEBUG     Delete: GuiBuildSettings
[2024-11-26 02:42:41,526]          guimain.py:964   DEBUG     Main focus switched to: GuiDocEditor
```

However, the font is reported in odt and docx as "Courier" with the correct font size; in the pdf it is "Courier New" with 1.5 pt font size.

**Comment by vkbo** [comment]
There is no info in the log if the font matches, as per the code snippet.

Courier and Courier New are selected by the matching algorithm based on internal logic, and is a special case. It's mentioned in the docs: https://doc.qt.io/qt-6/qfont.html#details

As for ODT output, the font is set explicitly as a string, so it's up to LibreOffice to do the matching. It probably does it differently. You can inspect what novelWriter generates by writing a .fodt file instead and look at the XML.

**Comment by vkbo** [comment]
> After changing "Application font", the wrong font name will still be displayed in the Build Settings. But this is only a display error and is probably covered by the advice to restart novelWriter after changing the Application font.

Missed this point. That's because the dialog shows the actual font, not the selected font, so it suffers from the same issue. However, the fontmatcher ensures that the correct font is used.

I could probably force the correct font match closer to the actual settings, and avoid these issues.

**Comment by HeyMyian** [comment]
> Courier and Courier New are selected by the matching algorithm based on internal logic, and is a special case. It's mentioned in the docs: https://doc.qt.io/qt-6/qfont.html#details

I see, this makes a lot of sense.
When writing a pdf from the `font_styles` branch with `Courier` selected as the font, the font size seems to be completely lost and it sets `1.5pt` instead for the whole pdf. As far as I can tell, this is new behavior in the branch; it works fine in 2.6 Beta 1.

**Comment by vkbo** [comment]
And you did not trigger my font matching code, which regenerates the font? (The log output)

**Comment by HeyMyian** [comment]
Ah, I see now how I can trigger the font matching. Yes, selecting Courier does **not** trigger the log output.

**Comment by vkbo** [comment]
Ok, just pushed some changes, Try now. But if it's not modifying the font, then I'm not sure what's happening to the PDF.

**Comment by vkbo** [comment]
Ok, so new version. This time, I'm forcing the printer resolution to 72 DPI, which should resolve fonts correctly, while the printer mode is still HighResolution (1200 DPI for PDF). They do mean different things apparently ... 🤷🏼‍♀️ 

In any event, Courier and Courier New now look alike, and I saw no issues on Linux.

**Comment by HeyMyian** [comment]
There is some regression when it comes to the kerning issues from #2100.

Current main branch:
✅ good kerning
❔ line spacing "1" in Build Settings, 1.16 in the pdf
![2024-11-26___20_17_33](https://github.com/user-attachments/assets/2343f885-11ca-4b2d-ad6d-70a43887af16)

This version of novelWriter:
https://github.com/vkbo/novelWriter/tree/07c61450c7a53c5185e25aae329504a6db09e843
❌ bad kerning
❔ line spacing "1" in Buils Settings, 1.33 in the pdf
![2024-11-26___20_14_34](https://github.com/user-attachments/assets/96e51df6-1e3c-4c68-921c-4d034832b3a0)

**Comment by vkbo** [comment]
It's not a regression. You cannot have your cake and eat it as far as I can tell here. Either the resolution is high, which prevents the kerning roundoff, or resolution is 72 DPI to support old fonts.

**Comment by vkbo** [comment]
The font database has a few checks for font scalability. They're listed in this area of the docs: https://doc.qt.io/qt-5/qfontdatabase.html#isBitmapScalable

My last attempt is to only use 1200 DPI for scalable fonts, otherwise use 72 DPI. With 72 DPI, the non-scalable fonts will be represented exactly, and I expect the kerning is adapted to 72 DPI scale.

You can try the branch again. If this doesn't work, then I will simply just drop support for bitmap fonts like Courier.

The curious thing is that it is still using Courier New, even if you selected Courier, so I don't know why it doesn't scale. But whatever.

**Comment by vkbo** [comment]
To summarise what's going on, and how it relates to the below point.

> ## 2.5.2
> 
>     * Setting the **Application font** overrides the Build Settings and puts the Application font name into the input field for Text Font in the MBS. Changing the MBS font is **not** possible. The preview will display the Application font.

By the way, I will not fix this in the 2.5.x branch. The issue here is, as already resolved, that Qt's font matching algorithm makes some weird choices when the application font changes. It seems to switch all other fonts to use it too. The documentation does mention the need to destroy and recreate all fonts in the app under certain circumstances, so I'm sure this is their solution. I.e. it's something like unintended behaviour, but not considered a bug per say.

The build setting dialog is designed to show the actual font used, rather then the font it claims to be, so it will show that despite the font selection, it will actually be using the application font. The dialog is correct, the font matching algorithm is the issue.

With my changes for 2.6 Beta 2, my font matcher will rectify the issue, including for the build settings. It will still show the actual font used rather than the requested font, but my font matcher will override the Qt behaviour (and issue a log warning when doing so).

This solves the issue for 2.6.

For 2.5 and before, the solution is to restart the application, which loads all fonts anew.

**Comment by HeyMyian** [comment]
I did a quick check of the branch, and I found all issues we discussed here are fixed 👍
